### ISSUE TITLE
Output Scorecard results as in-toto attestation

### ISSUE THREAD

**Issue by adityasaky** [ossf/scorecard#3352]
In addition to existing formats, it'd be handy to have scorecards result emitted as an in-toto attestation. The attestation framework is described here: https://github.com/in-toto/attestation.

## Why?

in-toto's attestation framework is a structured way to emit software supply chain specific claims. These claims can then be consumed by a policy engine like an in-toto "layout". Having scorecard results as an attestation would allow for its use alongside other attestations like SLSA provenance, test results, SCAI, as well as some popular SBOM formats like SPDX and CDX.

## What would it look like?

The document would have a lightweight in-toto Statement and the scorecard results themselves would be within it as a [predicate](https://github.com/in-toto/attestation/blob/main/spec/v1/predicate.md). Each predicate is identified using a type which is a URI. I think using the URL for this repository + version would work, but I'll leave it to the scorecard maintainers to decide!

## Open Questions

in-toto attestations are typically cryptographically signed and embedded in a signature envelope like DSSE. I'm not yet sure if scorecard supports that or intends to. By enabling that, the policy engine can also assert that the results were generated by a trusted entity (such as the correct CI workflow). Curious to hear your thoughts!

**Comment by adityasaky** [comment]
Whoops!

I'll spread the word a bit in the in-toto community to see if someone's interested in making this happen.

**Comment by bobcallaway** [comment]
@spencerschrock @laurentsimon @kpk47 curious as to your thoughts on this

**Comment by spencerschrock** [comment]
I agree with Ian that we do a similar thing already with the scorecard attestor code, although it does slightly different things (cc @raghavkaul). We also do something similar already with `ossf/scorecard-action`, where we use `hashrekord` log entries to accept crowd-sourced scores for the API. 

action side:
https://github.com/ossf/scorecard-action/blob/c64aac8aa26f9701232d21f7382097ea7044cd8d/signing/signing.go#L81-L82
webapp side:
https://github.com/ossf/scorecard-webapp/blob/28b861a9b9712b3a097ecf94e1057f3a7e802668/app/server/post_results.go#L133-L148

**Comment by raghavkaul** [comment]
Today the attestor subsystem attests whether a repo passes a YAML-based [policy](https://github.com/ossf/scorecard/tree/main/attestor#configuring-policies-for-scorecard-attestor), but maybe if scorecard signed/attested to the entire result output, that would give the score consumer more information about a repo.

**Comment by marcelamelara** [comment]
Pinging @zachariahcox, @TomHennen and @meder on this re: potential connection to the SLSA source/dependency track.

**Comment by meder** [comment]
Thanks, @marcelamelara.

One question I had was whether each scorecard check should have its own predicate type to enable composability where different tools can generate the same type of attestations and consumer is able to distinguish between different attesters. 

Has this been discussed somewhere?

**Comment by adityasaky** [comment]
I don't know if it's been discussed elsewhere, but I think as a starting point a single predicate type might suffice. I'm unaware of the different ways scorecard is deployed: are there cases where a subset of checks are used?

**Comment by spencerschrock** [comment]
> I don't know if it's been discussed elsewhere, but I think as a starting point a single predicate type might suffice. I'm unaware of the different ways scorecard is deployed: are there cases where a subset of checks are used?

The Scorecard binary supports all checks, but depending on who's running it the results may cover a different subset. Our API currently can return either 19 or 16 checks due to limitations we have running the analyses at scale. Additionally we have another result format now, so would this involve at least 4 predicate types?

1. The full Scorecard output (JSON)
2. Output for a single check (JSON)
3. full Scorecard output, but with "probes" (discrete heuristics which make up the check) instead of checks (JSON, but different structure)
4. Output of a single probe (see 3)

**Comment by ramonpetgrave64** [comment]
Another great thing to have is the ability to attest to scorecard results over a time period, or to verify that certain checks were not violated in previous x months or x releases. It seems like this could be possible for the Scorecard App with webhooks for Github Apps.
 - https://docs.github.com/en/webhooks/webhook-events-and-payloads#repository_ruleset

**Comment by spencerschrock** [comment]
> I think it might be quite easy for Scorecards to issue a VSA attesting to SLSA Source Level 2

I worry about Scorecard needing to produce different intoto predicates types to support some arbitrary number of external requirements that consumers may want. Which is partly why I suggested a Scorecard specific predicate in https://github.com/ossf/scorecard/issues/3352#issuecomment-2359417591 stating we observed some behavior(s). Consumers who want to build policy on top of that can consume our VSA and make a derivative VSA. (Is this a thing?)

**Comment by puerco** [comment]
OK, here is a PR with a change to add `--format=statement` and a predicate suggestion based on the current JSON output:

https://github.com/ossf/scorecard/pull/4491
